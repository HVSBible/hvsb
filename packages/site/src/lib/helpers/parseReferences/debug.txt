<p><strong>I FIGURED IT OUT. IF YOU PUT A COLON AFTER A VERSE REFERENCE IT DOUBLES THE TEXT. I TYPED A COMPLETE COPY AND EXPLORED A LOT OF DIFFERENT OPTIONS AND IT CAME OUT RIGHT IN THE FILE: THE ADORNING OF THE LAMB'S BRIDE</strong></p><p>The Lamb (Jesus) is glorified by the multitude because the wife of Jesus (the church) had prepared for the wedding. The bride's garment of \"bright, pure, and fine linen\" represents the righteous acts of the saints (Eph 5:23, 32). The practice of the Middle East in these times:</p><blockquote><p>\"The adorning of <u>the bride</u> was a very costly and elaborate affair. Much time was given to the preparation of her person. Every effort was put forth to make her complexion glossy and shining with a luster like unto marble. The words of David must have been their ideal for her: “that our daugh­ters may be as corner stones, polished after the similitude of a palace.” Psa. 144:12. Her dark locks of hair were often braided with gold and pearls. She was decked with all the precious stones and jewels that the family had inherited from previous generations. Those who were too poor to afford much would borrow what they could from their friends.\"[1] Fred H. Wight, <a href=\"https://www.amazon.com/Manners-Customs-Bible-Lands-Wight/dp/0802451756\">Manners and Customs of Bible Lands</a>, 130-131.</p></blockquote><p>This Middle Eastern custom is referenced in Jer 2:32: \"Can a maid forget her ornaments, or a bride her attire?\" and the words of the apostle John when he saw the New Jerusalem come down from God out of heaven in the new heavens and new earth, saying that the city was \"prepared as a bride adorned for her husband\" (Rev 21:2).</p>

From https://hvsb.app/WEB/REV/19/doc/8mz60eEDTvmmjtRLtsAT