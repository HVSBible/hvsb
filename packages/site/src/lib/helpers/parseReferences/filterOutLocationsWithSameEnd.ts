import type { Location } from './parseReferences'

export function filterOutLocationsWithSameEnd(locations: Location[]) {
  const alreadyFoundEndIndexes: number[] = [];
  return locations.filter((location) => {
    if (alreadyFoundEndIndexes.includes(location.end)) return false;
    alreadyFoundEndIndexes.push(location.end); return true;
  });
}

if (import.meta.vitest) {
  test('filterOutLocationsWithSameEnd', () => {
    const locations: Location[] = [
      {
        start: 8,
        end: 15,
        text: 'Genesis',
      },
      {
        start: 52,
        end: 58,
        text: '1 John',
      },
      {
        start: 54,
        end: 58,
        text: 'John',
      },
    ]
    expect(filterOutLocationsWithSameEnd(locations)).toEqual([
      {
        start: 8,
        end: 15,
        text: 'Genesis',
      },
      {
        start: 52,
        end: 58,
        text: '1 John',
      },
    ]);
  });
}
