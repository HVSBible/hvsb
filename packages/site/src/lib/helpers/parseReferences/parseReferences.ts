import { arrayOfBookNamesAbbreviations, getBookId } from "$lib/parts/data/books";
import { filterOutLocationsWithSameEnd } from "./filterOutLocationsWithSameEnd";

// NOTES:
// check to see if Dr. House uses Philemon 12 or Phileman 1:12, Jude 20 or Jude 1:20 to refer to verses in books with only 1 chapter
// watch out to see if we need to convert 'en' dash: `–` or 'em' dash: `—` to a hyphen with `.replace(/\u2013|\u2014/g, "-")`
// split a text into paragraphs by newline to keep book location memory from leaking over into next paragraph

export interface Reference {
  bookId?: string;
  chapter: number;
  verseRange: string; // 1 | 5-7 | 4a | 3-4a
  start: number;
  end: number;
  text: string;
}

export interface Location {
  start: number;
  end: number;
  text: string;
}

export function findReferencesInParagraph(string: string): Reference[] {
  const references: Reference[] = [];
  const bookNameLocations = getBookNameLocations(string);

  for (const [index, bookLocation] of bookNameLocations.entries()) {
    const nextLocation = bookNameLocations[index + 1];
    const endOfTextInfluencedByThisBookLocation = nextLocation?.start || string.length;

    const textInfluencedByThisBookLocation = string.slice(bookLocation.start, endOfTextInfluencedByThisBookLocation);

    const bookName = bookLocation.text;

    const chapterVerseReferencesFromThisBookLocation = findChapterVerseReferences(textInfluencedByThisBookLocation, bookName);

    const bookId = getBookId(bookName);
    for (const chapterVerseReference of chapterVerseReferencesFromThisBookLocation) {
      const reference: Reference = {
        bookId,
        ...chapterVerseReference,
        start: bookLocation.start + chapterVerseReference.start,
        end: bookLocation.start + chapterVerseReference.end,
      }
      references.push(reference);
    }
  }

  return references;
}

const bookNamesAndAbbreviations = arrayOfBookNamesAbbreviations();
const longestToShortestBookNames = bookNamesAndAbbreviations.sort((a, b) => b.length - a.length);
// check for book abbreviations used in actual articles and pull additional ones like (I John vs 1 John) from https://github.com/phillipb/bible-verse-parser/blob/master/lib/ValidBookNames.ts if helpful

export function getBookNameLocations(string: string): Location[] {
  const locations: Location[] = []
  // Find 1 John before John
  for (const name of longestToShortestBookNames) {
    const bookNameLocations = getBookIndexes(string, name);
    locations.push(...bookNameLocations);
  }
  const uniqueBookNameLocations = filterOutLocationsWithSameEnd(locations);
  const sortedByStartIndex = uniqueBookNameLocations.sort((a, b) => a.start - b.start);
  return sortedByStartIndex;
}

export function findChapterVerseReferences(string: string, bookName: string): Reference[] {
  const references: Reference[] = [];
  const chapterLocations = getChapterIndexes(string, bookName);

  for (const [chapterIndex, chapterLocation] of chapterLocations.entries()) {
    const chapter = parseInt(chapterLocation.text);

    const nextLocation = chapterLocations[chapterIndex + 1];
    const endOfTextInfluencedByThisChapterLocation = nextLocation?.start || string.length;
    const textInfluencedByThisChapterLocation = string.slice(chapterLocation.end, endOfTextInfluencedByThisChapterLocation);

    const verseLocations = getVerseRangeIndexes(textInfluencedByThisChapterLocation);

    for (const [verseIndex, verseLocation] of verseLocations.entries()) {
      const isFirstReferenceForBook = chapterIndex === 0 && verseIndex === 0;
      const isFirstReferenceForChapter = verseIndex === 0;

      const verseStart = isFirstReferenceForBook ? 0 :
        isFirstReferenceForChapter ? chapterLocation.start :
          chapterLocation.end + verseLocation.start;

      const verseEnd = chapterLocation.end + verseLocation.end;

      references.push({
        verseRange: verseLocation.text,
        chapter,
        start: verseStart,
        end: verseEnd,
        text: string.slice(verseStart, verseEnd),
      });
    };

    const chapterReferenceWithoutVerses = !verseLocations.length;
    if (chapterReferenceWithoutVerses) {
      const bookStart = 0;
      references.push({
        verseRange: "1",
        chapter,
        start: bookStart,
        end: chapterLocation.end,
        text: string.slice(bookStart, chapterLocation.end),
      });
    }
  }
  return references;
}

export function getBookIndexes(string: string, bookName: string): Location[] {
  const caseInsensitiveWholeWordBookName = new RegExp(`\\b(${bookName})\\b`, 'ig');
  return getIndexes(string, caseInsensitiveWholeWordBookName);
}

const NUMBERS_FOLLOWED_BY_COLON = /(\d+):/g;

export function getChapterIndexes(string: string, bookName: string): Location[] {
  const locations: Location[] = []

  const BOOK_OPTIONAL_PERIOD_FOLLOWED_BY_CHAPTER = new RegExp(`(?<=^${bookName}\\.?\\s)(\\d+)`, 'g'); // using positive lookbehind to ensure right after bookName without actually including those characters in the match index
  const initialChapter = getIndexes(string, BOOK_OPTIONAL_PERIOD_FOLLOWED_BY_CHAPTER);
  locations.push(...initialChapter);

  const chaptersWithVerses = getIndexes(string, NUMBERS_FOLLOWED_BY_COLON);
  chaptersWithVerses.forEach(location => {
    if (initialChapter[0]?.start !== location.start) locations.push(location)
  })

  return locations;
}

const LETTER_AT_BEGINNING_OF_WORD = /\b[a-zA-Z]/;
const BRACKET_OR_PERIOD = /[[.]/;
const combinedPattern = new RegExp(`${LETTER_AT_BEGINNING_OF_WORD.source}|${BRACKET_OR_PERIOD.source}`);

const NOT_PRECEEDED_BY_TEXT_OR_HYPHEN = /(?<!\w\s*-?)/;
const CAPTURE_NUMBER_OPTIONALLY_FOLLOWED_BY_NUMBERS_HYPHEN_OR_LETTERS = /(\b[0-9]+[a-zA-Z0-9-]*)/;
const NUMBERS_NOT_PRECEEDED_BY_TEXT_BUT_OPTIONALLY_FOLLOWED_BY_NUMBERS_HYPHEN_OR_LETTERS = new RegExp(`${NOT_PRECEEDED_BY_TEXT_OR_HYPHEN.source}${CAPTURE_NUMBER_OPTIONALLY_FOLLOWED_BY_NUMBERS_HYPHEN_OR_LETTERS.source}`, 'g');

export function getVerseRangeIndexes(string: string): Location[] {
  const charactersBeforeFirstWord = string.split(combinedPattern)[0] || string;
  return getIndexes(charactersBeforeFirstWord, NUMBERS_NOT_PRECEEDED_BY_TEXT_BUT_OPTIONALLY_FOLLOWED_BY_NUMBERS_HYPHEN_OR_LETTERS);
}

function getIndexes(string: string, regex: RegExp): Location[] {
  const matches: Location[] = [];
  for (const match of string.matchAll(regex)) {
    const text = match[1];
    matches.push({
      start: match.index,
      end: match.index + text.length,
      text,
    });
  }
  return matches;
}