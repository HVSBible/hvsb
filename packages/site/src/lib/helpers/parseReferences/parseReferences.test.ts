import { findChapterVerseReferences, findReferencesInParagraph, getBookIndexes, getBookNameLocations, getChapterIndexes, getVerseRangeIndexes } from './parseReferences';

describe('findReferencesInParagraph', () => {
  test('parses simple reference by itself', () => {
    expect(findReferencesInParagraph('Genesis 1:1')).toEqual([{
      bookId: 'GEN',
      chapter: 1,
      verseRange: '1',
      start: 0,
      end: 11,
      text: 'Genesis 1:1',
    }]);
  });

  test('returns first verse if book name is followed immediately by chapter number', () => {
    expect(findReferencesInParagraph('We see in this in Genesis 2.')).toEqual([{
      bookId: 'GEN',
      chapter: 2,
      start: 18,
      end: 27,
      text: 'Genesis 2',
    }]);
    expect(findReferencesInParagraph('We see in this in Gen. 2.')).toEqual([{
      bookId: 'GEN',
      chapter: 2,
      start: 18,
      end: 24,
      text: 'Gen. 2',
    }]);
  });

  test('does not parse ranges spanning chapters', () => {
    expect(findReferencesInParagraph('Genesis 1:1 - Genesis 2:1')).toEqual([
      {
        bookId: 'GEN',
        chapter: 1,
        end: 11,
        start: 0,
        text: 'Genesis 1:1',
        verseRange: '1',
      },
      {
        bookId: 'GEN',
        chapter: 2,
        end: 25,
        start: 14,
        text: 'Genesis 2:1',
        verseRange: '1',
      },
    ]);
  });

  test('handles a sentence full of references, separated by commas', () => {
    expect(findReferencesInParagraph('Gen. 5:12, Exod. 4:10-23, John 1:1-2, 3:1-2, 5:7-8')).toMatchSnapshot();
  });

  test('handles hyphen, en dash, and em dash the same', () => {
    const hyphenResult = findReferencesInParagraph('Gen 5:7-8');
    const enDashResult = findReferencesInParagraph('Gen 5:7–8');
    const emDashResult = findReferencesInParagraph('Gen 5:7—8');
    expect(hyphenResult).toEqual(enDashResult);
    expect(hyphenResult).toEqual(emDashResult);
  });

  test('handles Matt 5:1-2, 3, 3:1-2, 2 John 1:3 without marking the "2" in 2 John as verse 2 of Matthew 3', () => {
    expect(findReferencesInParagraph('Matt 5:1-2, 3, 3:1-2, 2 John 1:3')).toMatchSnapshot();
  });

  test('handles a conversational prose sentence with references', () => {
    expect(findReferencesInParagraph('If you go and read Gen 3:1-2, and 5:7-8, 12 then you will see 12 disciples.')).toEqual([
      {
        'bookId': 'GEN',
        'chapter': 3,
        'end': 28,
        'start': 19,
        'text': 'Gen 3:1-2',
        'verseRange': '1-2',
      },
      {
        'bookId': 'GEN',
        'chapter': 5,
        'end': 39,
        'start': 34,
        'text': '5:7-8',
        'verseRange': '7-8',
      },
      {
        'bookId': 'GEN',
        'chapter': 5,
        'end': 43,
        'start': 41,
        'text': '12',
        'verseRange': '12',
      },
    ]);
  });

  test('handles external links', () => {
    const linkWithClass = `Hello Genesis 1:2. Look at <a href="https://foo.com" class="hover:text-primary-800 underline" target="_blank">https://foo.com</a> to learn. Then turn to Exod. 2.`
    expect(findReferencesInParagraph(linkWithClass)).toMatchSnapshot();

    const urlWithNumbers = `<p>God promised David (2 Samuel 7:12-13).THE <a href="https://www.amazon.com/s?k=h+wayne+house&amp;crid=55ULNGFAPAI3&amp;sprefix=%2Caps%2C405&amp;ref=nb_sb_ss_i_1_0">REJECTION OF JEHOIACHIN</a> (Coniah, Jeconiah)`
    expect(findReferencesInParagraph(urlWithNumbers)).toMatchSnapshot();
  })

  test('First chapter match must follow right after book name', () => {
    const firstJohnProblem = `John tells us something (12:9). This caused Lazarus (12:10).`
    expect(findReferencesInParagraph(firstJohnProblem)).toEqual([]);
  })

  test('Ignores trailing hyphen cause by redundant chapter usage or range between chapters', () => {
    const redundantChapter = `Acts 10:1-10:48`
    expect(findReferencesInParagraph(redundantChapter)).toEqual([
      {
        'bookId': 'ACT',
        'chapter': 10,
        'end': 9,
        'start': 0,
        'text': 'Acts 10:1',
        'verseRange': '1',
      },
      {
        'bookId': 'ACT',
        'chapter': 10,
        'end': 15,
        'start': 10,
        'text': '10:48',
        'verseRange': '48',
      },
    ]);
    const chapterRange = `Acts 10:1-11:20`
    expect(findReferencesInParagraph(chapterRange)).toEqual([
      {
        'bookId': 'ACT',
        'chapter': 10,
        'end': 9,
        'start': 0,
        'text': 'Acts 10:1',
        'verseRange': '1',
      },
      {
        'bookId': 'ACT',
        'chapter': 11,
        'end': 15,
        'start': 10,
        'text': '11:20',
        'verseRange': '20',
      },
    ]);
  })

  test.fails('ignores references inside links', () => {
    const referenceInsideLink = `important article "<a href="https://media.foo.com">Viticulture and John 15:1-6</a>" (<i>Bibliotheca Sacra</i>)`
    expect(findReferencesInParagraph(referenceInsideLink)).toEqual([]);
  })

  test.fails('trailing colon to explain something following a reference', () => {
    expect(findReferencesInParagraph('Remember what we see in Luke 16:10: 20 men! Then we also find in 17:5: 40 men!')).toEqual([{
      bookId: 'LUK',
      chapter: 16,
      verseRange: '10',
      start: 24,
      end: 34,
      text: 'Luke 16:10',
    }]);

    expect(findReferencesInParagraph('important article on John 15:2-6: "Viticulture"...')).toEqual([{
      'bookId': 'JHN',
      'chapter': 15,
      'start': 21,
      'end': 32,
      'text': 'John 15:2-6',
      'verseRange': '2-6',
    }]);
  });
});

test('getBookNameLocations', () => {
  expect(getBookNameLocations('This is Genesis 1:1 and Exod 2:5. Then we reference 1 John 2:1.')).toEqual([
    {
      start: 8,
      end: 15,
      text: 'Genesis',
    },
    {
      start: 24,
      end: 28,
      text: 'Exod',
    },
    {
      start: 52,
      end: 58,
      text: '1 John',
    },
  ]);
});

describe('findChapterVerseReferences', () => {
  test('basic', () => {
    expect(findChapterVerseReferences('Genesis 1:1', 'Genesis')).toEqual([{
      text: 'Genesis 1:1',
      chapter: 1,
      verseRange: '1',
      start: 0,
      end: 11,
    }]);
  });

  test('chapter with multiple references', () => {
    expect(findChapterVerseReferences('John 5:1-2, 12', 'John')).toMatchInlineSnapshot(`
      [
        {
          "chapter": 5,
          "end": 10,
          "start": 0,
          "text": "John 5:1-2",
          "verseRange": "1-2",
        },
        {
          "chapter": 5,
          "end": 14,
          "start": 12,
          "text": "12",
          "verseRange": "12",
        },
      ]
    `);
  });

  test('chapter with multiple references', () => {
    expect(findChapterVerseReferences('John 3:1, 5:1-2, 12', 'John')).toMatchInlineSnapshot(`
      [
        {
          "chapter": 3,
          "end": 8,
          "start": 0,
          "text": "John 3:1",
          "verseRange": "1",
        },
        {
          "chapter": 5,
          "end": 15,
          "start": 10,
          "text": "5:1-2",
          "verseRange": "1-2",
        },
        {
          "chapter": 5,
          "end": 19,
          "start": 17,
          "text": "12",
          "verseRange": "12",
        },
      ]
    `);
  });

  test('sentence with three references', () => {
    expect(findChapterVerseReferences('Gen. 1:1, 2:3, and 4:5-7 and some more here and else.', 'Gen')).toEqual([
      {
        text: 'Gen. 1:1',
        chapter: 1,
        verseRange: '1',
        start: 0,
        end: 8,
      },
      {
        text: '2:3',
        chapter: 2,
        verseRange: '3',
        start: 10,
        end: 13,
      },
      {
        text: '4:5-7',
        chapter: 4,
        verseRange: '5-7',
        start: 19,
        end: 24,
      },
    ]);
  });

  test('verseRange is undefined if no verse specified', () => {
    expect(findChapterVerseReferences('Genesis 1', 'Genesis')).toEqual([{
      text: 'Genesis 1',
      chapter: 1,
      verseRange: undefined,
      start: 0,
      end: 9,
    }]);
  });

  test('books with only 1 chapter', () => {
    const resultWithChapter = findChapterVerseReferences('2 John 1:4', '2 John');
    const resultWithOutChapter = findChapterVerseReferences('2 John 4', '2 John');
    expect(resultWithChapter[0].verseRange).toEqual(resultWithOutChapter[0].verseRange);

    expect(findChapterVerseReferences('Phlm 6', 'Phlm')).toEqual([{
      text: 'Phlm 6',
      chapter: 1,
      verseRange: '6',
      start: 0,
      end: 6,
    }]);
  });
});

describe('getBookIndexes', () => {
  test('return start, end, and original text', () => {
    expect(getBookIndexes('Gen 1:1', 'Gen')).toEqual([{
      start: 0,
      end: 3,
      text: 'Gen',
    }]);
    expect(getBookIndexes('This is Genesis 1:1', 'Genesis')).toEqual([{
      start: 8,
      end: 15,
      text: 'Genesis',
    }]);
  });
  test('handles period after abbreviated book name', () => {
    expect(getBookIndexes('Matt. 5:2', 'Matt')).toEqual([{
      start: 0,
      end: 4,
      text: 'Matt',
    }]);
  });
});

describe('getChapterIndexes', () => {
  test('handles chapter and verse anywhere', () => {
    expect(getChapterIndexes('Gen. 2:3 and 14:2, 5', 'Gen')).toEqual([
      {
        'start': 5,
        'end': 6,
        'text': '2',
      },
      {
        'start': 13,
        'end': 15,
        'text': '14',
      },
    ]);
  });
  test('handles just chapter immediately at beginning, after space or period and space', () => {
    expect(getChapterIndexes('Gen 2 is a place to read...', 'Gen')).toEqual([
      {
        'start': 4,
        'end': 5,
        'text': '2',
      },
    ]);
    expect(getChapterIndexes('Gen. 2.', 'Gen')).toEqual([
      {
        'start': 5,
        'end': 6,
        'text': '2',
      },
    ]);
  });
  test('does not pay attention to random number not at beginning', () => {
    expect(getChapterIndexes(' well talk about something with 2 pigons', 'Exodus')).toEqual([]);
  });
});

describe('getVerseRangeIndexes', () => {
  test('captures verse number at beginning of string', () => {
    expect(getVerseRangeIndexes('3 and ')).toEqual([
      {
        'end': 1,
        'start': 0,
        'text': '3',
      },
    ]);
  });

  test('handles ranges, and multiple verses following commas or semicolons', () => {
    expect(getVerseRangeIndexes('3a-5, 7; 12, ')).toEqual([
      {
        'end': 4,
        'start': 0,
        'text': '3a-5',
      },
      {
        'end': 7,
        'start': 6,
        'text': '7',
      },
      {
        'end': 11,
        'start': 9,
        'text': '12',
      },
    ]);
  });

  test('does not capture number in brackets', () => {
    const numberInsideBrackets = `:41 [3]`
    expect(getVerseRangeIndexes(numberInsideBrackets)).toEqual([
      {
        'start': 1,
        'end': 3,
        'text': '41',
      },
    ]);
  });

  test('does not get verse indexes for numbers coming after a period', () => {
    const numberAfterMedialPeriod = `:41). 3 sheep`
    expect(getVerseRangeIndexes(numberAfterMedialPeriod)).toEqual([
      {
        'start': 1,
        'end': 3,
        'text': '41',
      },
    ]);
  });

  test('does not get verse indexes for numbers coming after a word', () => {
    expect(getVerseRangeIndexes('3 and 12 sheep.')).toEqual([
      {
        'end': 1,
        'start': 0,
        'text': '3',
      },
    ]);
    expect(getVerseRangeIndexes(':2. Look at <a href="https://foo.com" class="hover:text-primary-800">https://foo.com</a> to learn')).toEqual(
      [
        {
          'end': 2,
          'start': 1,
          'text': '2',
        },
      ]);
    expect(getVerseRangeIndexes(':12-13).THE <a href="https://www.amazon.com/s?k=h+wayne+house&amp;crid=55ULNGFAPAI3&amp;sprefix=%2Caps%2C405&amp;ref=nb_sb_ss_i_1_0">REJECTION OF JEHOIACHIN</a> (Coniah, Jeconiah)')).toEqual([{
      'end': 6,
      'start': 1,
      'text': '12-13',
    }])
  })
});
