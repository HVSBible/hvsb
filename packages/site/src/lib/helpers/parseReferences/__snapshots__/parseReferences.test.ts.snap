// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findReferencesInParagraph > handles Matt 5:1-2, 3, 3:1-2, 2 John 1:3 without marking the "2" in 2 John as verse 2 of Matthew 3 1`] = `
[
  {
    "bookId": "MAT",
    "chapter": 5,
    "end": 10,
    "start": 0,
    "text": "Matt 5:1-2",
    "verseRange": "1-2",
  },
  {
    "bookId": "MAT",
    "chapter": 5,
    "end": 13,
    "start": 12,
    "text": "3",
    "verseRange": "3",
  },
  {
    "bookId": "MAT",
    "chapter": 3,
    "end": 20,
    "start": 15,
    "text": "3:1-2",
    "verseRange": "1-2",
  },
  {
    "bookId": "2JN",
    "chapter": 1,
    "end": 32,
    "start": 22,
    "text": "2 John 1:3",
    "verseRange": "3",
  },
]
`;

exports[`findReferencesInParagraph > handles a sentence full of references separated by commas 1`] = `
[
  {
    "bookId": "GEN",
    "chapter": 5,
    "end": 9,
    "start": 0,
    "text": "Gen. 5:12",
    "verseRange": "12",
  },
  {
    "bookId": "EXO",
    "chapter": 4,
    "end": 24,
    "start": 11,
    "text": "Exod. 4:10-23",
    "verseRange": "10-23",
  },
  {
    "bookId": "JHN",
    "chapter": 1,
    "end": 36,
    "start": 26,
    "text": "John 1:1-2",
    "verseRange": "1-2",
  },
  {
    "bookId": "JHN",
    "chapter": 3,
    "end": 43,
    "start": 38,
    "text": "3:1-2",
    "verseRange": "1-2",
  },
  {
    "bookId": "JHN",
    "chapter": 5,
    "end": 50,
    "start": 45,
    "text": "5:7-8",
    "verseRange": "7-8",
  },
]
`;

exports[`findReferencesInParagraph > handles external links 1`] = `
[
  {
    "bookId": "GEN",
    "chapter": 1,
    "end": 17,
    "start": 6,
    "text": "Genesis 1:2",
    "verseRange": "2",
  },
  {
    "bookId": "EXO",
    "chapter": 2,
    "end": 160,
    "start": 153,
    "text": "Exod. 2",
    "verseRange": "1",
  },
]
`;

exports[`findReferencesInParagraph > handles external links 2`] = `
[
  {
    "bookId": "2SA",
    "chapter": 7,
    "end": 39,
    "start": 23,
    "text": "2 Samuel 7:12-13",
    "verseRange": "12-13",
  },
]
`;
